convex/lp/simplex.go

```diff
@@ -0,0 +1,601 @@
+// solves that can be improved by doing rank-one updates. For example, the swap
+// step is just a rank-one update.
+// TODO(btracey): Better handling on the linear solve errors. If the condition
+// number is not inf and the equation solved "well", should keep moving.
+
+var (
+	ErrBland      = errors.New("lp: bland: all replacements are negative or cause ill-conditioned ab")
+	ErrInfeasible = errors.New("lp: problem is infeasible")
+	ErrLinSolve   = errors.New("lp: linear solve failure")
+	ErrUnbounded  = errors.New("lp: problem is unbounded")
+	ErrSingular   = errors.New("lp: A is singular")
+	ErrZeroColumn = errors.New("lp: A has a column of all zeros")
+	ErrZeroRow    = errors.New("lp: A has a row of all zeros")
+)
+
+var (
+	badShape = "lp: size mismatch"
+)
+
+// TODO(btracey): Should these tolerances be part of a settings struct?
+
+const (
+	// initPosTol is the tolerance on the initial condition being feasible. Strictly,
+	// the x should be positive, but instead it must be greater than -initPosTol.
+	initPosTol = 1e-13
+	// blandNegTol is the tolerance on the value being greater than 0 in the bland test.
+	blandNegTol = 1e-14
+	// rRoundTol is the tolerance for rounding values to zero when testing if
+	// constraints are met.
+	rRoundTol = 1e-13
+	// dRoundTol is the tolerance for testing if values are zero for the problem
+	// being unbounded.
+	dRoundTol = 1e-13
+	// phaseIZeroTol tests if the Phase I problem returned a feasible solution.
+	phaseIZeroTol = 1e-12
+	// blandZeroTol is the tolerance on testing if the bland solution can move.
+	blandZeroTol = 1e-12
+)
+
+// Simplex solves a linear program in standard form. The standard form of a linear
+// program is:
+//  minimize	c^T x
+//  s.t. 		A*x = b
+//  			x >= 0 .
+// The input tol sets how close to the optimal solution is found (specifically,
+// when the maximal reduced cost is below tol). An error will be returned if the
+// problem is infeasible or unbounded. In rare cases, numeric errors can cause
+// the Simplex to fail. In this case, an error will be returned along with the
+// most recently found feasible solution.
+//
+// The Convert function can be used to transform a general LP into standard form.
+//
+// The input matrix A must have full rank and may not contain any columns with
+// all zeros. Furthermore, len(c) must equal the number of columns of A, and len(b)
+// must equal the number of rows of A. Simplex will panic if these conditions are
+// not met.
+//
+// initialBasic can be used to set the initial set of indices for a feasible
+// solution to the LP. If an initial feasible solution is not known, initialBasic
+// may be nil. If initialBasic is non-nil, len(initialBasic) must equal the number
+// of columns of A and must be an actual feasible solution to the LP, otherwise
+// Simplex will panic.
+//
+// For a detailed introduction to the Simplex method please see lectures 11-13 of
```

Comment:
Wrong simplex method :). This is Danzig's Simplex, not [Nelder-Mead](https://godoc.org/github.com/gonum/optimize#NelderMead)